feat: add sha checksum to binaries

hugorut opened the pull request:
Adds checksums to release artefacts.

![Screenshot 2022-02-08 at 14 14 04](https://user-images.githubusercontent.com/6455139/153004613-929393da-a004-4a61-be98-8a05b4119848.png)

We then recommend users check these as part of install process, as outlined here: https://github.com/infracost/docs/pull/161

hugorut commented on scripts/install.sh:
```diff
@@ -4,10 +4,39 @@ set -e
 os=$(uname | tr '[:upper:]' '[:lower:]')
 arch=$(uname -m | tr '[:upper:]' '[:lower:]' | sed -e s/x86_64/amd64/)
+if [ "$arch" = "aarch64" ]; then
+  arch="arm64"
```
solves issue where `arrch64` architecture machines can't get a valid binary

hugorut commented on scripts/install.sh:
```diff
@@ -4,10 +4,39 @@ set -e
 echo
 echo "Moving /tmp/infracost-$os-$arch to /usr/local/bin/infracost (you might be asked for your password due to sudo)"
-sudo mv /tmp/infracost-$os-$arch /usr/local/bin/infracost
+if [ -x "$(command -v sudo)" ]; then
+  sudo mv /tmp/infracost-$os-$arch /usr/local/bin/infracost
+else
+  mv /tmp/infracost-$os-$arch /usr/local/bin/infracost
+fi
```
I've removed the reliance on `sudo` as when I was testing this in `ubuntu` container I didn't actually have access to it. Just mv the binary to the local bin was fine perms wise.

hugorut commented:
@aliscott could you check the `install.sh` changes here: https://github.com/infracost/infracost/pull/1351/commits/ee868919e91cf5f51bf3d1d6eb4e877dfd8f874e as we discussed yesterday.

aliscott commented:
LGTM!

alikhajeh1 commented on scripts/install.sh:
```diff
@@ -4,10 +4,39 @@ set -e
+    (cd /tmp/;shasum -sqc $tar.sha256)
+    rm /tmp/$tar.sha256
+  else
+    echo "The shasum command could not be found, skipping checksum validation"
```
I'm wondering if we should avoid logging here and on line 19 if there's a 404. In both cases, there's nothing the user can do (since the install proceeds anyway), so why warn them? The one we want to flag-up is when the checksum fails. What does that log look like? I'd even consider not logging if the checksum passes.

I say this based on the `brew install` experience, it's so painless/unalarming that I didn't even know it was checking the checksum.

hugorut replied:
I don't have a strong opinion of this - however, I remember seeing a brew issue (can't find it at the moment) where someone mentioned they wanted to have a warning if the cask had disabled `sha256`. Which suggests that we might want to be as verbose as possible. 

The warning message when checksum fails is as follows:

```
infracost-darwin-amd64.tar.gz: FAILED
shasum: WARNING: 1 computed checksum did NOT match
```

This is directly from the shasum command. We could capture output and modify message if you want.

alikhajeh1 replied:
I'd optimize this flow for getting to success as soon as possible. The 404 log doesn't have any actionable things for the user to do, so I'd probably skip that so we can merge this PR. The missing shasum command gives me a bit of pause - maybe "Skipping checksum validation as the shasum command could not be found, no action needed." is clearer that we don't want the user to do anything, this is just FYI.

Re the failure log, maybe it could say "Installation checksum failed! This could be a security issue so please email hello@infracost.io for help."
For now, I guess it's ok to continue with the installation even if the checksum failed until we revisit this side of things next month.

alikhajeh1 commented on scripts/install.sh:
```diff
@@ -4,10 +4,39 @@ set -e
+fi
+
+tar xzf /tmp/$tar -C /tmp
+rm /tmp/$tar
```
Should we get rid of the `rm` commands? it could alarm users if the `rm` asks for confirmation, and we're deleting stuff in /tmp so it doesn't matter as it'll get cleaned up eventually if storage is low.

hugorut replied:
Personally I'd prefer scripts to cleanup artefacts they created. It just seems good practice.

alikhajeh1 replied:
tru dat! Overall it seems like we want this script to be non-interactive as people are copying that curl into their workflows, so if there is no risk of `rm` throwing a prompt then it's safe to ship.

alikhajeh1 reviewed (commented):
Good stuff!

alikhajeh1 commented:
it might be worth checking https://www.shellcheck.net (I use to use that on the old diff script). Also, I suppose we should merge this after the next release as the docs repo uses `curl -fsSL https://raw.githubusercontent.com/infracost/infracost/master/scripts/install.sh | sh` so if it goes into master it gets used.

hugorut commented:
> Also, I suppose we should merge this after the next release as the docs repo uses `curl -fsSL https://raw.githubusercontent.com/infracost/infracost/master/scripts/install.sh | sh` so if it goes into master it gets used. 

This is safe to go into master at the moment - hence `404` checking. But we can wait until the release.

hugorut commented:
@alikhajeh1 I've updated the script following your feedback, could you please take a look at it again:

1. updated to fail silently on 404
2. updated shasum not found messaging
3. updated failure messaging if shasum check fails
4. fixed shell linting quotes

**if shasum fails:**
<img width="789" alt="Screenshot 2022-02-10 at 09 19 47" src="https://user-images.githubusercontent.com/6455139/153377270-173289ef-054b-4976-b8a1-45536dc98674.png">

**If shasum succeeds:**
<img width="857" alt="Screenshot 2022-02-10 at 09 20 10" src="https://user-images.githubusercontent.com/6455139/153377284-24cda2a4-432f-405b-9527-1a5e4b430e04.png">

**if 404:**
<img width="883" alt="Screenshot 2022-02-10 at 09 20 26" src="https://user-images.githubusercontent.com/6455139/153377296-81c9690e-476f-4d4f-ae6a-fbf444aec542.png">

**If shasum not exists:**
<img width="839" alt="Screenshot 2022-02-10 at 09 26 52" src="https://user-images.githubusercontent.com/6455139/153377581-63fed659-1d0f-4937-8dad-0155de0dbfbb.png">

---


> For now, I guess it's ok to continue with the installation even if the checksum failed until we revisit this side of things next month. 

I'm hesitant to do this as I think failure of checksum should be force stop for users as this could be a critical security issue. So we explicitly exit if the shasum fails. @aliscott what are your thoughts?

aliscott commented:
Here's my thoughts, but these aren't strong opinions.

1. Skipping checksum validation because of 404

I think we should show a similar message here as the `shasum could not be found` error. It seems inconsistent not to and I think we should be explicit.

2. Failure on mismatch checksum

One option would be to have a prompt like 'The checksum does not match. Would you like to continue?'. If we think most users would use this in CLI instead of CI then I think having one is fine, we potentially prompt anyway for `sudo`.

hugorut commented:
@aliscott I've added the 404 check back in with https://github.com/infracost/infracost/pull/1351/commits/b653e35ffb4360b8ff4683bbef616113c3d9a415 and added a prompt for shasum failure
